src/Http/GenericRouter.php

```diff
@@ -172,4 +160,68 @@ private function createResponse(Response|View $input): Response
 
         return $input;
     }
+
+    private function matchStaticRoute(Request $request): ?MatchedRoute
+    {
+        $staticRoute = $this->getRoutes()[$request->method->value][$request->getPath()] ?? null;
+
+        if ($staticRoute === null) {
+            return null;
+        }
+
+        // TODO: why do we need to check this?
+        if ($staticRoute->isDynamic) {
+            return null;
+        }
+
+        return new MatchedRoute($staticRoute, []);
+    }
+
+    private function matchDynamicRoute(Request $request): ?MatchedRoute
+    {
+        // If there are no routes for the given request method, we immediately stop
+        $routesForMethod = $this->getRoutes()[$request->method->value] ?? null;
+        if ($routesForMethod === null) {
+            return null;
+        }
+
+        // Next, we'll build one big regex to match the correct route
+        // See https://github.com/tempestphp/tempest-framework/pull/175 for the details
+
+        /** @var \Tempest\Http\Route[] $routesForMethod */
+        $routesForMethod = array_values($routesForMethod);
+        $combinedMatchingRegex = "#^(?|";
+
+        foreach ($routesForMethod as $routeIndex => $route) {
+            if (! $route->isDynamic) {
+                continue; // TODO: why this check?
+            }
+
+            $combinedMatchingRegex .= "$route->matchingRegex (*" . self::MARK_TOKEN . ":$routeIndex) |";
+        }
+
+        $combinedMatchingRegex = rtrim($combinedMatchingRegex, '|');
+        $combinedMatchingRegex .= ')$#x';
+
+        // Then we'll use this regex to see whether we have a match or not
+        $matchResult = preg_match($combinedMatchingRegex, $request->getPath(), $matches);
+
+        if (! $matchResult || ! array_key_exists(self::MARK_TOKEN, $matches)) {
+            return null;
+        }
+
+        $route = $routesForMethod[$matches[self::MARK_TOKEN]];
+
+        // TODO: we could probably optimize resolveParams now,
+        // because we already know for sure there's a match
+        $routeParams = $this->resolveParams($route, $request->getPath());
+
+        // This check should _in theory_ not be needed,
+        // since we're certain there's a match
+        if ($routeParams === null) {
```

Comment:
This check might be unnecessary if we refactor `resolveParams`

Reply:
True! The only case where it would return null right now is if the route Regex does not match which we have ensured before.